pkg/api/version/version.go

```diff
@@ -0,0 +1,257 @@
+	logger   log.Logger
+	cancel   context.CancelFunc
+	ctx      context.Context
+	addr, id string
+}
+
+func New(cfg util.CommonRingConfig, logger log.Logger, reg prometheus.Registerer) (*Service, error) {
+	client, err := kv.NewClient(cfg.KVStore, GetCodec(), kv.RegistererWithKVName(reg, "versions"), logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize versions' KV store")
+	}
+
+	instanceAddr, err := ring.GetInstanceAddr(cfg.InstanceAddr, cfg.InstanceInterfaceNames, logger, false)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	instancePort := ring.GetInstancePort(cfg.InstancePort, cfg.ListenPort)
+	svc := &Service{
+		store:  client,
+		id:     cfg.InstanceID,
+		addr:   fmt.Sprintf("%s:%d", instanceAddr, instancePort),
+		cfg:    cfg,
+		logger: log.With(logger, "component", "versions"),
+		cancel: cancel,
+		ctx:    ctx,
+	}
+	// The service is simple only has a running function.
+	// Stopping is manual to ensure we stop as part of the shutdown process.
+	svc.BasicService = services.NewBasicService(
+		func(_ context.Context) error { return nil },
+		svc.running,
+		func(_ error) error { return nil },
+	)
+	return svc, nil
+}
+
+func (svc *Service) running(ctx context.Context) error {
+	go svc.loop()
+	<-ctx.Done()
+	return nil
+}
+
+func (svc *Service) loop() {
+	ticker := time.NewTicker(15 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := svc.heartbeat(svc.ctx); err != nil {
+				level.Error(svc.logger).Log("msg", "failed to heartbeat", "err", err)
+			}
+
+		case <-svc.ctx.Done():
+			level.Info(svc.logger).Log("msg", "versions is shutting down")
+			return
+		}
+	}
+}
+
+func (svc *Service) heartbeat(ctx context.Context) error {
+	return svc.store.CAS(ctx, "versions", func(in interface{}) (out interface{}, retry bool, err error) {
+		var versions *versionv1.Versions
+		if in == nil {
+			versions = newVersions().(*Versions).Versions
+		} else {
+			versions = in.(*Versions).Versions
+		}
+		current, ok := versions.Instances[svc.id]
+		if !ok {
+			current = &versionv1.InstanceVersion{}
+			versions.Instances[svc.id] = current
+		}
+		current.Addr = svc.addr
+		current.ID = svc.id
+		current.Timestamp = time.Now().Unix()
+		current.QuerierAPI = currentQuerierVersion
+		// Now prune old instances.
+		for id, instance := range versions.Instances {
+			lastHeartbeat := time.Unix(instance.GetTimestamp(), 0)
+
+			if time.Since(lastHeartbeat) > 1*time.Minute {
+				level.Warn(svc.logger).Log("msg", "auto-forgetting instance from the ring because it is unhealthy for a long time", "instance", id, "last_heartbeat", lastHeartbeat.String())
+				delete(versions.Instances, id)
+			}
+		}
+		return &Versions{
+			Versions: versions,
+		}, true, nil
+	})
+}
+
+func (svc *Service) Version(ctx context.Context, req *connect.Request[versionv1.VersionRequest]) (*connect.Response[versionv1.VersionResponse], error) {
```

Comment:
The service can be passed to other components and allow them to also verify the version of the cluster too.

For example when adding new feature to only store-gateways.